File: src/dataset/c/protein_translation_Exercism_m3g4d1v3r/openAI/ChatGPT4_protein_translation_v1.c
#include "protein_translation.h"
#include <string.h>

int translate(proteins_t *data, codon_protein_t *map, const char *rna) {
    if (strlen(rna) < 3) {
        data->valid = (*rna == '\0');
        return -1;
    }

    char codon[4] = {rna[0], rna[1], rna[2], '\0'};

    for (protein_t prot = 0; prot <= Stop; prot++) {
        for (size_t idx = 0; map[prot].codons[idx][0] != '\0'; idx++) {
            if (strcmp(map[prot].codons[idx], codon) == 0) {
                if (prot == Stop) {
                    data->valid = true;
                    return 0;
                }
                data->proteins[data->count++] = prot;
                return 0;
            }
        }
    }
    data->valid = false;
    return -1;
}

proteins_t proteins(const char *const rna) {
    proteins_t data = {.count = 0, .valid = false};
    codon_protein_t map[] = {
        {Methionine, {"AUG", ""}},
        {Phenylalanine, {"UUU", "UUC", ""}},
        {Leucine, {"UUA", "UUG", ""}},
        {Serine, {"UCU", "UCC", "UCA", "UCG", ""}},
        {Tyrosine, {"UAU", "UAC", ""}},
        {Cysteine, {"UGU", "UGC", ""}},
        {Tryptophan, {"UGG", ""}},
        {Stop, {"UAA", "UAG", "UGA", ""}}
    };

    const char *rna_ptr = rna;
    while (!data.valid && translate(&data, map, rna_ptr) == 0) {
        rna_ptr += 3;
    }
    return data;
}

File: src/dataset/go/protein_translation_Exercism_go_ThomasZumsteg/openAI/ChatGPT4_protein_translation_v1.go
package protein

var codonMap = map[string]string{
	"AUG": "Methionine",
	"UUU": "Phenylalanine", "UUC": "Phenylalanine",
	"UUA": "Leucine", "UUG": "Leucine",
	"UCU": "Serine", "UCC": "Serine", "UCA": "Serine", "UCG": "Serine",
	"UAU": "Tyrosine", "UAC": "Tyrosine",
	"UGU": "Cysteine", "UGC": "Cysteine",
	"UGG": "Tryptophan",
	"UAA": "STOP", "UAG": "STOP", "UGA": "STOP",
}

func FromCodon(codon string) string {
	if protein, exists := codonMap[codon]; exists {
		return protein
	}
	panic("Not a valid codon: " + codon)
}

func FromRNA(rna string) (codons []string) {
	for i := 0; i+3 <= len(rna); i += 3 {
		codon := FromCodon(rna[i : i+3])
		if codon == "STOP" {
			break
		}
		codons = append(codons, codon)
	}
	return
}